File: ios/ExpoSquircleView.swift
import UIKit
import ExpoModulesCore
import PocketSVG

class ExpoSquircleView: ExpoView {
    let squircleLayer = CAShapeLayer()
    var radius: CGFloat = 0
    var cornerSmoothing: CGFloat = 0
    var preserveSmoothing: Bool = false
    var isAnimationEnabled: Bool = false
    
    required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        setupSquircleLayer()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupSquircleLayer() {
        layer.addSublayer(squircleLayer)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        squircleLayer.frame = bounds
        squircleLayer.path = createSquirclePath()
    }
    
    private func createSquirclePath() -> CGPath {
        let width: CGFloat = bounds.width
        let height: CGFloat = bounds.height
        
        let path = SquirclePath.create(width: width - squircleLayer.lineWidth, height: height - squircleLayer.lineWidth, radius: radius, cornerSmoothing: cornerSmoothing, preserveSmoothing: preserveSmoothing);
        
        // if borderWidth is greater than 0, we need to shift the shape
        // to match the original width & height
        var translationTransform = CGAffineTransform(translationX: squircleLayer.lineWidth / 2, y: squircleLayer.lineWidth / 2)
        
        if let translatedPath = path.copy(using: &translationTransform) {
            return translatedPath
        }
        
        return path
    }
    
    func setBackgroundColor(_ backgroundColor: UIColor) {
        CATransaction.begin()
        CATransaction.setDisableActions(!isAnimationEnabled)
        squircleLayer.fillColor = backgroundColor.cgColor
        CATransaction.commit()
    }
    
    func setBorderColor(_ borderColor: UIColor) {
        CATransaction.begin()
        CATransaction.setDisableActions(!isAnimationEnabled)
        squircleLayer.strokeColor = borderColor.cgColor
        CATransaction.commit()
    }
    
    func setBorderWidth(_ borderRadius: CGFloat) {
        CATransaction.begin()
        CATransaction.setDisableActions(!isAnimationEnabled)
        squircleLayer.lineWidth = borderRadius
        CATransaction.commit()
        setNeedsLayout()
    }
    
    func setCornerSmoothing(_ cornerSmoothing: CGFloat) {
        self.cornerSmoothing = cornerSmoothing
        squircleLayer.setNeedsLayout()
        setNeedsLayout()
    }
    
    func setRadius(_ radius: CGFloat) {
        self.radius = radius
        squircleLayer.setNeedsLayout()
        setNeedsLayout()
    }
    
    func setPreserveSmoothing(_ preserveSmoothing: Bool) {
        self.preserveSmoothing = preserveSmoothing
        squircleLayer.setNeedsLayout()
        setNeedsLayout()
    }
}

File: android/src/main/java/expo/modules/squircleview/ExpoSquircleView.kt
package expo.modules.squircleview

import SquirclePath
import android.content.Context
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.Path
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView

class ExpoSquircleView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val borderPaint = Paint(paint)
    private val path = Path()
    private var cornerSmoothing = 0
    private var borderColor = 0xFF000000.toInt()
    private var borderWidth = 0f
    private var backgroundColor = 0x00000000
    private var borderRadius = 0f
    private var preserveSmoothing = false

    init {
        paint.color = backgroundColor
        paint.style = Paint.Style.FILL
        setWillNotDraw(false)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        paint.color = backgroundColor
        canvas.drawPath(path, paint)

        if (borderWidth > 0) {
            borderPaint.color = borderColor
            borderPaint.style = Paint.Style.STROKE
            borderPaint.strokeWidth = Utils.convertDpToPixel(borderWidth, context).toFloat()
            canvas.drawPath(path, borderPaint)
        }
    }

    override fun onSizeChanged(newWidth: Int, newHeight: Int, oldWidth: Int, oldHeight: Int) {
        super.onSizeChanged(newWidth, newHeight, oldWidth, oldHeight)
        resetSquirclePath(newWidth.toFloat(), newHeight.toFloat())
    }

    private fun resetSquirclePath(
        width: Float,
        height: Float
    ) {
        if (width == 0f || height == 0f) {
            return
        }

        val pixelBorderWidth = Utils.convertDpToPixel(this.borderWidth, context);

        val squirclePath = SquirclePath(
            width - pixelBorderWidth,
            height - pixelBorderWidth,
            borderRadius,
            cornerSmoothing / 100f,
            preserveSmoothing
        )

        // if borderWidth is greater than 0, we need to shift the shape
        // to match the original width & height
        val shiftX = pixelBorderWidth / 2f
        val shiftY = pixelBorderWidth / 2f
        val translationMatrix = Matrix().apply {
            setTranslate(shiftX, shiftY)
        }
        val translatedPath = Path().apply {
            squirclePath.transform(translationMatrix, this)
        }

        path.reset()
        path.addPath(translatedPath)
    }

    fun setCornerSmoothing(c: Int) {
        cornerSmoothing = c
        resetSquirclePath(width.toFloat(), height.toFloat())
        invalidate()
    }

    fun setBorderRadius(b: Float) {
        val pixelRadius = Utils.convertDpToPixel(b, context)
        borderRadius = pixelRadius

        resetSquirclePath(width.toFloat(), height.toFloat())
        invalidate()
    }

    fun setPreserveSmoothing(p: Boolean) {
        preserveSmoothing = p

        resetSquirclePath(width.toFloat(), height.toFloat())
        invalidate()
    }

    fun setViewBackgroundColor(color: Int) {
        backgroundColor = color
        paint.color = backgroundColor
        invalidate()
    }

    fun setBorderColor(color: Int) {
        borderColor = color
        invalidate()
    }

    fun setBorderWidth(width: Float) {
        borderWidth = width
        resetSquirclePath(this.width.toFloat(), this.height.toFloat())
        invalidate()
    }
}
